### PR TITLE
JIT: Produce less convoluted IR for boolean `isinst` checks

### LINKED ISSUE #36649: Unoptimal codegen for "obj is T" with T being struct/sealed
Follow up from a question in https://github.com/dotnet/runtime/pull/1817 ([here](https://github.com/dotnet/runtime/pull/1817#issuecomment-629714632)), cc. @EgorBo.

### Description

I think I've identified 4 scenarios where the JIT doesn't produce optimal codegen for an `object is T` or `object is T variable` expression, when `T` is either a `struct` or a `sealed class.

<details>
  <summary><b>object is T, when T is a struct</b> (click to expand)</summary>
<br/>

```csharp
public static bool Is_Slow<T>(object obj) where T : struct
{
    return obj is T;
}
```

```asm
; using T = int
C.Is_Slow[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: test rcx, rcx
    L0003: je short L0016
    L0005: mov rax, 0x7ff9d6bdb1e8
    L000f: cmp [rcx], rax
    L0012: je short L0016
    L0014: xor ecx, ecx
    L0016: test rcx, rcx
    L0019: setne al
    L001c: movzx eax, al
    L001f: ret
```

Note how the JIT creates two separate branches, one per condition (`null` check and type check). This can be improved by just rewriting the code manually to perform those two checks individually:

```csharp
public static bool Is_Fast<T>(object obj) where T : struct
{
    return obj != null && obj.GetType() == typeof(T);
}
```

```asm
C.Is_Fast[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: test rcx, rcx
    L0003: je short L0019
    L0005: mov rax, 0x7ff9d6bdb1e8
    L000f: cmp [rcx], rax
    L0012: sete al
    L0015: movzx eax, al
    L0018: ret
    L0019: xor eax, eax
    L001b: ret
```

Here the type check is just done with a `cmp` + `setz`, removing one conditional branch entirely.

</details>

<details>
  <summary><b>object is T value, when T is a struct</b> (click to expand)</summary>
<br/>

```csharp
public static T UnboxOrDefault_Slow<T>(object obj) where T : struct
{
    return (obj is T value) ? value : default;
}
```

```asm
C.UnboxOrDefault_Slow[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: push rsi
    L0001: sub rsp, 0x20
    L0005: mov rsi, rcx
    L0008: mov rdx, rsi
    L000b: test rdx, rdx
    L000e: je short L0021
    L0010: mov rcx, 0x7ff9d6bdb1e8
    L001a: cmp [rdx], rcx
    L001d: je short L0021
    L001f: xor edx, edx
    L0021: test rdx, rdx
    L0024: je short L0050
    L0026: mov rdx, 0x7ff9d6bdb1e8
    L0030: cmp [rsi], rdx
    L0033: je short L0047
    L0035: mov rdx, rsi
    L0038: mov rcx, 0x7ff9d6bdb1e8
    L0042: call 0x00007ffa366e04a0
    L0047: mov eax, [rsi+8]
    L004a: add rsp, 0x20
    L004e: pop rsi
    L004f: ret
    L0050: xor eax, eax
    L0052: add rsp, 0x20
    L0056: pop rsi
    L0057: ret
```

In this case the JIT creates 4 branches, two for the `is` check and 2 for the `unbox.any` opcode, as the runtime unfortunately still doesn't support/emit the `no.` prefix. Anyway, here's with explicit code:

```csharp
public static T UnboxOrDefault_Fast<T>(object obj) where T : struct
{
    return obj != null && obj.GetType() == typeof(T) ? (T)obj : default;
}
```

```asm
C.UnboxOrDefault_Fast[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: push rsi
    L0001: sub rsp, 0x20
    L0005: mov rsi, rcx
    L0008: test rsi, rsi
    L000b: je short L001c
    L000d: mov rax, 0x7ff9d6bdb1e8
    L0017: cmp [rsi], rax
    L001a: je short L0024
    L001c: xor eax, eax
    L001e: add rsp, 0x20
    L0022: pop rsi
    L0023: ret
    L0024: mov rdx, 0x7ff9d6bdb1e8
    L002e: cmp [rsi], rdx
    L0031: je short L0045
    L0033: mov rdx, rsi
    L0036: mov rcx, 0x7ff9d6bdb1e8
    L0040: call 0x00007ffa366e04a0
    L0045: mov eax, [rsi+8]
    L0048: add rsp, 0x20
    L004c: pop rsi
    L004d: ret
```

As with the previous case, one less conditional branch and slightly smaller codegen.

</details>

<details>
  <summary><b>object is T, when T is a sealed class</b> (click to expand)</summary>
<br/>

```csharp
public sealed class Model
{
    public static bool Is_Slow<T>(object obj)
    {
        return obj is Model;
    }
}
```

```asm
Model.Is_Slow[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: test rcx, rcx
    L0003: je short L0016
    L0005: mov rax, 0x7ff9ded6cdf0
    L000f: cmp [rcx], rax
    L0012: je short L0016
    L0014: xor ecx, ecx
    L0016: test rcx, rcx
    L0019: setne al
    L001c: movzx eax, al
    L001f: ret
```

And here is with the manual checks just like the first two cases:

```csharp
public sealed class Model
{
    public static bool Is_Fast<T>(object obj)
    {
        return obj != null && obj.GetType() == typeof(Model);
    }
}
```

```asm
Model.Is_Fast[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: test rcx, rcx
    L0003: je short L0019
    L0005: mov rax, 0x7ff9ded6cdf0
    L000f: cmp [rcx], rax
    L0012: sete al
    L0015: movzx eax, al
    L0018: ret
    L0019: xor eax, eax
    L001b: ret
```

</details>

<details>
  <summary><b>object is T value, when T is a sealed class</b> (click to expand)</summary>
<br/>

```csharp
public sealed class Model
{
    public static Model GetOrNull_Slow<T>(object obj)
    {
        if (obj is Model model) return model;
        return null;
    }
}
```

```asm
Model.GetOrNull_Slow[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: mov rax, rcx
    L0003: test rax, rax
    L0006: je short L0019
    L0008: mov rdx, 0x7ff9df11ce10
    L0012: cmp [rax], rdx
    L0015: je short L0019
    L0017: xor eax, eax
    L0019: test rax, rax
    L001c: je short L001f
    L001e: ret
    L001f: xor eax, eax
    L0021: ret
```

As above, one redundant conditional branch. Here is with explicit checks, note I'm using `Unsafe.As<T>(object)` here to force the JIT not to emit additional checks, as a standard `(T)` cast would result in worse codegen.

```csharp
public sealed class Model
{
    public static Model GetOrNull_Fast<T>(object obj)
    {
        if (obj != null && obj.GetType() == typeof(Model))
        {
            return Unsafe.As<Model>(obj);
        }
        return null;
    }
}
```

```asm
Model.GetOrNull_Fast[[System.Int32, System.Private.CoreLib]](System.Object)
    L0000: test rcx, rcx
    L0003: je short L0018
    L0005: mov rax, 0x7ff9df11ce10
    L000f: cmp [rcx], rax
    L0012: jne short L0018
    L0014: mov rax, rcx
    L0017: ret
    L0018: xor eax, eax
    L001a: ret
```

Here we once again have one less conditional branch than the one produced by the `is` operator.

> **Note:** in this last case, we could rewrite the first method as simply returning `as Model`, which correctly optimizes the final codegen and results in even smaller code size. I figured it was still worth pointing out the missed optimization when writing the code through the `is` operator though, as devs might very well still use it for a variety of reasons.

</details>

There are mainly two potential improvements I'm seeing:
- One less conditional branch in the "fast" version
- Slightly smaller codegen (this might in part go away if the method is inlined though)

### Configuration

Tested on [sharplab.io](https://sharplab.io/), in `Default`, `x64` and Roslyn `master` branches.
All assembly is from the Release configuration.

category:cq
theme:optimization
skill-level:expert
cost:medium
impact:small

### PR DESCRIPTION
**Pull request by jakobbotsch**
Currently the IR for boolean `isinst` checks ends up being something like `(typecheck(x) ? x : null) != null`, which the JIT ends up having a hard time clean up early. With object stack allocation this pattern usually leads to unnecessary address exposure.

This adds a simple pattern match during import to produce less convoluted IR in the common cases where the `isinst` is just used as a boolean check.

Example from #36649:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static bool Is_Slow(object obj) =>
    obj is int;
```

```diff
 ; Method C:Is_Slow(System.Object):ubyte (FullOpts)
 G_M38459_IG01:  ;; offset=0x0000
 						;; size=0 bbWeight=1 PerfScore 0.00
 
 G_M38459_IG02:  ;; offset=0x0000
        test     rcx, rcx
        je       SHORT G_M38459_IG05
 						;; size=5 bbWeight=1 PerfScore 1.25
 
 G_M38459_IG03:  ;; offset=0x0005
        mov      rax, 0x7FFB89FA1C00      ; System.Int32
        cmp      qword ptr [rcx], rax
        jne      SHORT G_M38459_IG05
 						;; size=15 bbWeight=0.25 PerfScore 1.06
 
 G_M38459_IG04:  ;; offset=0x0014
+       mov      eax, 1
        jmp      SHORT G_M38459_IG06
-						;; size=2 bbWeight=0.12 PerfScore 0.25
+						;; size=7 bbWeight=0.12 PerfScore 0.28
 
-G_M38459_IG05:  ;; offset=0x0016
-       xor      rcx, rcx
+G_M38459_IG05:  ;; offset=0x001B
+       xor      eax, eax
 						;; size=2 bbWeight=0.25 PerfScore 0.06
 
-G_M38459_IG06:  ;; offset=0x0018
-       test     rcx, rcx
-       setne    al
-       movzx    rax, al
-						;; size=9 bbWeight=1 PerfScore 1.50
-
-G_M38459_IG07:  ;; offset=0x0021
+G_M38459_IG06:  ;; offset=0x001D
        ret      
 						;; size=1 bbWeight=1 PerfScore 1.00
-; Total bytes of code: 34
+; Total bytes of code: 30

```

Fix #36649

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -5464,6 +5464,32 @@ GenTree* Compiler::impOptimizeCastClassOrIsInst(GenTree* op1, CORINFO_RESOLVED_T
...
+        case CEE_BRFALSE_S:
+            *consumed = 0;
+            return true;
+        case CEE_BRTRUE:
+        case CEE_BRTRUE_S:
+            *consumed = 0;
+            return true;
+        case CEE_LDNULL:
+            nextOpcode = impGetNonPrefixOpcode(codeAddr + 1, codeEndp);
+            if (nextOpcode == CEE_CGT_UN)
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:5482]: can be CEQ as well? not sure if it's popular

**Comment by EgorBo** [comment]
Should close https://github.com/dotnet/runtime/issues/36649

File: src/coreclr/jit/importer.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp:5482]: I had it initially, but it requires reversing the condition and I didn't manage to get Roslyn to ever produce it, so I decided to get rid of it to simplify the logic. I just ran SPMI replay and only saw a handful of test-only hits, so I think I just won't bother. Looks like `impBoxPatternMatch` doesn't handle it either.

**Comment by jakobbotsch** [comment]
> Should close #36649

Looks like it, your `Is_Slow` example from there:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
public static bool Is_Slow(object obj) =>
    obj is int;
```

Base:
```asm
; Method C:Is_Slow(System.Object):ubyte (FullOpts)
G_M38459_IG01:  ;; offset=0x0000
						;; size=0 bbWeight=1 PerfScore 0.00

G_M38459_IG02:  ;; offset=0x0000
       test     rcx, rcx
       je       SHORT G_M38459_IG05
						;; size=5 bbWeight=1 PerfScore 1.25

G_M38459_IG03:  ;; offset=0x0005
       mov      rax, 0x7FFB89FA1C00      ; System.Int32
       cmp      qword ptr [rcx], rax
       jne      SHORT G_M38459_IG05
						;; size=15 bbWeight=0.25 PerfScore 1.06

G_M38459_IG04:  ;; offset=0x0014
       jmp      SHORT G_M38459_IG06
						;; size=2 bbWeight=0.12 PerfScore 0.25

G_M38459_IG05:  ;; offset=0x0016
       xor      rcx, rcx
						;; size=2 bbWeight=0.25 PerfScore 0.06

G_M38459_IG06:  ;; offset=0x0018
       test     rcx, rcx
       setne    al
       movzx    rax, al
						;; size=9 bbWeight=1 PerfScore 1.50

G_M38459_IG07:  ;; offset=0x0021
       ret      
						;; size=1 bbWeight=1 PerfScore 1.00
; Total bytes of code: 34
```

Diff:
```asm
; Method C:Is_Slow(System.Object):ubyte (FullOpts)
G_M38459_IG01:  ;; offset=0x0000
						;; size=0 bbWeight=1 PerfScore 0.00

G_M38459_IG02:  ;; offset=0x0000
       test     rcx, rcx
       je       SHORT G_M38459_IG05
						;; size=5 bbWeight=1 PerfScore 1.25

G_M38459_IG03:  ;; offset=0x0005
       mov      rax, 0x7FFB89FC1C00      ; System.Int32
       cmp      qword ptr [rcx], rax
       je       SHORT G_M38459_IG05
						;; size=15 bbWeight=0.25 PerfScore 1.06

G_M38459_IG04:  ;; offset=0x0014
       mov      eax, 1
       jmp      SHORT G_M38459_IG06
						;; size=7 bbWeight=0.12 PerfScore 0.28

G_M38459_IG05:  ;; offset=0x001B
       xor      eax, eax
						;; size=2 bbWeight=0.25 PerfScore 0.06

G_M38459_IG06:  ;; offset=0x001D
       ret      
						;; size=1 bbWeight=1 PerfScore 1.00
; Total bytes of code: 30
```

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -5464,6 +5464,55 @@ GenTree* Compiler::impOptimizeCastClassOrIsInst(GenTree* op1, CORINFO_RESOLVED_T
...
+    {
+        case CEE_BRFALSE:
+        case CEE_BRFALSE_S:
+            // BRFALSE/BRTRUE importation are expected to transparently handle
+            // that the created tree is a TYP_INT instead of TYP_REF, so we do
+            // not consume them here.
+            *consumed = 0;
+            return true;
+        case CEE_BRTRUE:
+        case CEE_BRTRUE_S:
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:5500]: nit:
```c
        case CEE_BRFALSE:
        case CEE_BRFALSE_S:
        case CEE_BRTRUE:
        case CEE_BRTRUE_S:
            *consumed = 0;
            return true;
```

**Comment by jakobbotsch** [comment]
Overall the IR created by this PR is not really any less convoluted than the existing IR. Instead of `(x != null ? x.mt == expectedMT ? x : null : null) != null`, we create `(x != null ? x.mt == expectedMT ? 1 : 0 : 0) != 0`.
I initially tried to create something like `(x != null ? x.mt == expectedMT : 0) != 0`, but that resulted in quite mixed diffs.
Even so, it still seems to help in some of the reported cases, and it will help for #103361 as well when `x` is stack allocated.

Of course it would be better if we could create the more natural control flow, but that's hard given the representation through `QMARK`s (when we are doing late expansion we create the more natural flow directly, but IIRC the early expansion here still turns out to be more beneficial).

**Comment by EgorBo** [comment]
> but IIRC @EgorBo left the early expansion here in place because it is still more beneficial).

I initially hoped we'd enable JitOptRepeat and I could move the late-cast expansion to the JitOptRepeat loop and the 2nd iteration of it would clean most of the regressions I hit just fine 😞

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @EgorBo @AndyAyersMS 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=709807&view=ms.vss-build-web.run-extensions-tab)

File: src/coreclr/jit/importer.cpp [hunk 3]
```diff
@@ -5578,19 +5634,37 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
...
     op1 = impCloneExpr(op1, &op1Clone, CHECK_SPILL_ALL, nullptr DEBUGARG("ISINST eval op1"));
 
-    GenTreeOp*    condMT    = gtNewOperNode(GT_NE, TYP_INT, gtNewMethodTableLookup(op1Clone), op2);
-    GenTreeOp*    condNull  = gtNewOperNode(GT_EQ, TYP_INT, gtClone(op1), gtNewNull());
-    GenTreeQmark* qmarkMT   = gtNewQmarkNode(TYP_REF, condMT, gtNewColonNode(TYP_REF, gtNewNull(), gtClone(op1)));
-    GenTreeQmark* qmarkNull = gtNewQmarkNode(TYP_REF, condNull, gtNewColonNode(TYP_REF, gtNewNull(), qmarkMT));
+    if (*booleanCheck)
+    {
+        GenTreeOp*    condMT   = gtNewOperNode(GT_NE, TYP_INT, gtNewMethodTableLookup(op1Clone), op2);
+        GenTreeOp*    condNull = gtNewOperNode(GT_EQ, TYP_INT, gtClone(op1), gtNewNull());
```
Comment (EgorBo) [inline src/coreclr/jit/importer.cpp:5640]: nit: looks like these two branches have a lot of common code, e.g. `condMT` and `condNull` are the same.

**Review by EgorBo — APPROVED** [review]
LGTM, minor clean up comment, probably, not worth re-running CI for

File: src/coreclr/jit/importer.cpp (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/importer.cpp]: Deduplicated a bit of it

**Pull request merged**